tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommand.java

```diff
@@ -330,13 +403,433 @@ Map<TopicPartition, OffsetsInfo> getOffsets(StreamsGroupDescription description)
+                        Function.identity(),
+                        topicPartition -> resetPlanForGroup.get(topicPartition).offset()));
+
+                    return checkOffsetsRange(requestedOffsets).entrySet().stream()
+                        .collect(Collectors.toMap(Map.Entry::getKey, e -> new OffsetAndMetadata(e.getValue())));
+                }).orElseGet(Collections::emptyMap);
+            } else if (opts.options.has(opts.resetToCurrentOpt)) {
+                Map<TopicPartition, OffsetAndMetadata> currentCommittedOffsets = getCommittedOffsets(groupId);
+                Collection<TopicPartition> partitionsToResetWithCommittedOffset = new ArrayList<>();
+                Collection<TopicPartition> partitionsToResetWithoutCommittedOffset = new ArrayList<>();
+
+                for (TopicPartition topicPartition : partitionsToReset) {
+                    if (currentCommittedOffsets.containsKey(topicPartition))
+                        partitionsToResetWithCommittedOffset.add(topicPartition);
+                    else
+                        partitionsToResetWithoutCommittedOffset.add(topicPartition);
+                }
+
+                Map<TopicPartition, OffsetAndMetadata> preparedOffsetsForPartitionsWithCommittedOffset = partitionsToResetWithCommittedOffset.stream()
+                    .collect(Collectors.toMap(Function.identity(), topicPartition -> {
+                        OffsetAndMetadata committedOffset = currentCommittedOffsets.get(topicPartition);
+
+                        if (committedOffset == null) {
+                            throw new IllegalStateException("Expected a valid current offset for topic partition: " + topicPartition);
+                        }
+
+                        return new OffsetAndMetadata(committedOffset.offset());
+                    }));
+
+                Map<TopicPartition, OffsetAndMetadata> preparedOffsetsForPartitionsWithoutCommittedOffset = getLogEndOffsets(partitionsToResetWithoutCommittedOffset)
+                    .entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> {
+                        if (!(e.getValue() instanceof LogOffset)) {
+                            CommandLineUtils.printUsageAndExit(opts.parser, "Error getting ending offset of topic partition: " + e.getKey());
+                        }
+
+                        return new OffsetAndMetadata(((LogOffset) e.getValue()).value);
+                    }));
+
+                preparedOffsetsForPartitionsWithCommittedOffset.putAll(preparedOffsetsForPartitionsWithoutCommittedOffset);
+
+                return preparedOffsetsForPartitionsWithCommittedOffset;
+            }
+
+            CommandLineUtils.printUsageAndExit(opts.parser, String.format("Option '%s' requires one of the following scenarios: %s", opts.resetOffsetsOpt, opts.allResetOffsetScenarioOpts));
+            return null;
+        }
+
+        Optional<Map<String, Map<TopicPartition, OffsetAndMetadata>>> resetPlanFromFile() {
```

Comment:
I assume quite a bit of this code was copied from the consumer group tool. Is there any chance of sharing some code? May not be the case, but since most of this code is just dealing with topic partitions and not with tasks, I wonder if there is an easy way to refactor functions like this into a common utility.